Signup: Add autofill for username when available. [feedback wanted]. Implements #753

SawyerHood opened the pull request:
This PR adds the feature in #753. In the signup flow, if the user has set a site name and registered it while signing up, upon reaching the account creation form the username field will already be filled in with the site name chosen. 

Here are images of the feature.
http://zippy.gfycat.com/SoulfulCourteousAnole.gif

I am fairly new to React so I do have a question, I feel like this code should be abstracted more and not be implemented SignupForm. Is there a better place to check if a site name has been checked? Thanks in advanced.

fabianapsimoes commented:
cc @scruffian @drewblaisdell

scruffian commented:
@SawyerHood - thanks for the pull request - I'm reviewing it now. Just a note to remind you that we have a <a href="https://github.com/Automattic/wp-calypso/blob/master/docs/git-workflow.md#branch-naming-scheme">branch naming scheme</a>.

scruffian commented on client/components/signup-form/index.jsx:
```diff
@@ -54,6 +56,24 @@ export default React.createClass( {
+		if ( ! this.props.signupProgressStore ) {
+			return form;
+		}
+		let siteStepProgress = find( this.props.signupProgressStore, function( step ) {
```
You _could_ use ES6 syntax here:
```
step => { return step.stepName === 'site'; }
```

scruffian replied:
Actually this could just be:

```step => step.stepName === site```

scruffian commented:
This works perfectly :)

I have a few thoughts about how to proceed with it:

1. It is potentially an annoyance to users to have the username field automatically populated; it would be good to a/b test this change to see whether it has an impact on conversions - we have an `abtest` lib that will help you with that

2. Not many flows use the `site-creation` step. It would be really nice to get this into the `domain` step too, although it probably makes sense to run an ab test on this variation first.

3. There are some cases where a site could be available but the username is taken. In these cases the user will see an error, which isn't terrible, but just something to be aware of.

Thanks again for the pull request; it's great :+1:

SawyerHood commented:
Thanks @scruffian for the feedback. I should have some time tonight or tomorrow to get to this.

1. I will defiantly look into the a/b test library and set up an a/b test.

2. If the A/B test works out well I can implement it with domain step.

3. It might be better to check if the username has been taken and if so, not autofill the field.

4. This shouldn't be too bad, I can take a look at it.

Once again thank you.

scruffian commented:
> It might be better to check if the username has been taken and if so, not autofill the field.

The cases of this happening will be pretty few, so I think we can try the test first...

SawyerHood commented:
Ok @scruffian, I finally had some time to get back to this PR. I generalized it into a utility function named autoFillField. It takes a configuration object with the state of the form to change, the progress store, the name of the field to autofill, the step name to auto-fill from in the progress store, and finally the property name to autofill-from in the progress store. I also enabled an abtest for the site flow with 50% of the users having it autofill. I also added a unit test for the autoFillField function.

## Test Process

### Auto Fill Enabled
1. log out and create a new account with the developer signup flow.
2. Modify local storage so that the autoFillUsernameSignup abtest is set to 'autoFill'
3. Proceed through the sign in progress
4. When you get to the email sign up form, the username should be auto-filled with the site name that you entered.

### Auto Fill Not Enabled
1. log out and create a new account with the developer signup flow.
2. Modify local storage so that the autoFillUsernameSignup abtest is set to 'dontAutoFill'
3. Proceed through the sign in progress
4. When you get to the email sign up form, the username should **NOT** be auto-filled with the site name that you entered.

I also tested this with the 'main' signup flow grabbing from the domain and it works like a charm, but I didn't include the code here. Also, I didn't see any flows that had the user step come before the site step. Did you have a particular one in mind?

drewblaisdell commented on client/signup/utils.js:
```diff
@@ -93,6 +95,25 @@ function getNextStepName( flowName, currentStepName ) {
 }
 
+
+function autoFillField( config ) {
```
Great idea to move this into its own function in `signup/utils`. A couple super minor notes:

- `autoFillField` is a sort of generic name for what this does. What do you think about `autoFillFieldWithProgressData` or something?
- If we decide to add this functionality to the domains step, the data we'll be using to auto-fill the username will need to be modified, as it will be a domain like `foobarbaz.net`. Splitting this into two functions (one to grab field x from step y and one to add it to the form) would give us the option of modifying the data before it is injected into the form. This also might make naming these functions easier. What do you think?
- _Super_ minor and not necessary to change, but you can include the destructured assignment right in the function signature if you want:

```
function autoFillField( { form, stepName, fieldToTake, fieldToMerge, signupProgress } )
```

drewblaisdell commented on client/signup/utils.js:
```diff
@@ -93,6 +95,25 @@ function getNextStepName( flowName, currentStepName ) {
+	if ( fieldToMergeValue === null && siteStepProgress ) {
+		let mergeObject = {};
+		mergeObject[ fieldToMerge ] = { value: siteStepProgress[ fieldToTake ] };
+		return merge( form, mergeObject );
```
Also unnecessary to change, but something you might find interesting nonetheless: in ES6, objects can be defined with computed keys, which can obviate `mergeObject` altogether:

```js
return merge( form, {
  [ fieldToMerge ]: { value: siteStepProgress[ fieldToTake ] }
} );
```

SawyerHood replied:
Now this is awesome! Thanks for pointing this out.

drewblaisdell commented on client/signup/utils.js:
```diff
@@ -93,6 +95,25 @@ function getNextStepName( flowName, currentStepName ) {
+	if ( ! signupProgressStore ) {
+		return form;
+	}
+	let siteStepProgress = find(
```
We try to use `const` instead of `let` for all identifiers we only assign to once (which, I believe, is all of the ones in this function).

drewblaisdell commented:
This looks great! I had a few minor notes.

> I also tested this with the 'main' signup flow grabbing from the domain and it works like a charm, but I didn't include the code here.

Awesome. We probably need this to work with data from the domains step in order to have enough participants in this test. As you can see in `signup/config/flows`, the only signup flows that use the `site` step are test flows or the developer flow. Would it be difficult to commit that code as well?

I wouldn't worry handling the case where the user step appears before the site/domain step until this has been a/b tested.

SawyerHood commented:
@drewblaisdell, thanks for the great code review. I updated my PR with your mention changes and also added auto-fill from the domain flow. 

While filling the username I do the following:
1. If the user uses a free wordpress.com domain just use the subdomain
2. If the user uses their own domain, remove the top level domain and use that
3. If the user uses their own domain with a subdomain, just use the subdomain.

The test process for this is similar to the original but while checking for all three of these cases.

drewblaisdell commented on client/signup/utils.js:
```diff
@@ -93,6 +95,27 @@ function getNextStepName( flowName, currentStepName ) {
+
+function mergeFormWithValue( { form, fieldName, fieldValue} ) {
+	let fieldToMergeValue = formState.getFieldValue( form, fieldName );
+	if ( fieldToMergeValue == null ) {
```
`fieldToMergeValue` can be declared with `const` because we only assign to it once, but you may as well just omit it with:

```js
if ( ! formState.getFieldValue( form, fieldName ) ) {
```

This will merge for all falsy values (including an empty string), which is different than the behavior above (which will merge for `undefined` and `null`). I think either behavior is probably fine, but if we want to check for null, we should do a strict `===` comparison so that we know we're merging fields that have been initialized and set to `null`.

drewblaisdell commented on client/components/signup-form/index.jsx:
```diff
@@ -55,6 +57,27 @@ export default React.createClass( {
 	},
 
+	autoFillUsername( form ) {
+		let siteName = getValueFromProgressStore( {
```
`let` should be `const` here because we only assign to `siteName` once.

Also, good idea to use a single object in the function signature for `getValueFromProgressStore` and `mergeFormWithValue`. This is much more readable than `getValueFromProgressStore( 'site', 'site', signupProgressStore );`

drewblaisdell commented on client/signup/test/utils-test.js:
```diff
@@ -124,3 +124,41 @@ describe( 'utils.getValidPath - currentFlowName === defaultFlowName', function()
 } );
 
+describe( 'utils.getValueFromProgressStore', function() {
+	let testStore = [ { stepName: 'empty' }, { stepName: 'site', site: 'calypso' } ];
```
All of the `let`s in this file should be `const`s.

drewblaisdell commented on client/signup/utils.js:
```diff
@@ -93,6 +95,27 @@ function getNextStepName( flowName, currentStepName ) {
 }
 
+function getValueFromProgressStore( { signupProgressStore, stepName, fieldName } ) {
+	if ( ! signupProgressStore ) {
```
I think we should move the blocks that check for the presence of the progress store out of these two utility functions and into `SignupForm`. The main reason for this is that it will make the behavior of `SignupForm` (which must behave with and without the `signupProgressStore` prop) more explicit if we skip `autoFillUsername` altogether when that component isn't provided `signupProgressStore`.

drewblaisdell commented on client/components/signup-form/index.jsx:
```diff
@@ -67,7 +90,11 @@ export default React.createClass( {
 		} );
-		this.setState( { form: this.formStateController.getInitialState() } );
+		let initialState = this.formStateController.getInitialState();
+		if ( abtest( 'autoFillUsernameSignup' ) === 'autoFill' ) {
```
`SignupForm` is mounted in the main signup at `/start` in the `user` step, but also elsewhere in Calypso when users accept an invite, in `LoggedOutInvite`. We only have the progress store in the former case, so we probably only want to assign them to the test if that is present. We can achieve this by checking for the `signupProgressStore` here:

```js
if ( this.props.signupProgressStore && abtest( 'autoFillUsernameSignup' ) === 'autoFill' ) {
```

This will also allow you to remove the checks for `! signupProgressStore` inside the utility functions.

drewblaisdell commented:
I had a few more minor notes, but otherwise, it's almost ready to merge. This works great, really nice work on this.

Some notes about the branch itself:

- This needs to be rebased with `master` to resolve conflicts with some changes in `SignupForm`.
- We should probably squash the first four commits in this PR.

drewblaisdell commented:
@SawyerHood: Sorry, I didn't realize that you rebased! If you go ahead and rebase again (there are conflicts again) and ping me here afterward, I'll more quickly assert that it was rebased properly and merge the PR. I would do this myself, but I don't have access to your fork.

Otherwise, I can cherry-pick your commit into a new branch and do the rebase myself there. You'll get credit for the contribution either way.

SawyerHood commented:
Hey @drewblaisdell, I went ahead and resolved the merge conflict. Thanks again.
